packages/wrangler/src/inspect.ts

```diff
@@ -0,0 +1,526 @@
+  // (stored in, no surprises here, `remoteWebSocket`)
+  useEffect(() => {
+    if (!props.inspectorUrl) {
+      return;
+    }
+    // The actual websocket instance
+    const ws = new WebSocket(props.inspectorUrl);
+    setRemoteWebSocket(ws);
+
+    /**
+     * A handle to the interval we run to keep the websocket alive
+     */
+    let keepAliveInterval: NodeJS.Timer;
+
+    /**
+     * Test if the websocket is closed
+     */
+    function isClosed() {
+      return (
+        ws.readyState === WebSocket.CLOSED ||
+        ws.readyState === WebSocket.CLOSING
+      );
+    }
+
+    /**
+     * Send a message to the remote websocket
+     */
+    function send(event: Record<string, unknown>): void {
+      if (!isClosed()) {
+        ws.send(JSON.stringify(event));
+      }
+    }
+
+    /**
+     * Closes the inspector.
+     */
+    function close(): void {
+      if (!isClosed()) {
+        try {
+          ws.close();
+        } catch (err) {
+          // Closing before the websocket is ready will throw an error.
+        }
+      }
+    }
+
+    /**
+     * Since we have a handle on the remote websocket, we can tap
+     * into its events, and log any pertinent ones directly to
+     * the terminal (which means you have insight into your worker
+     * without having to open the devtools).
+     */
+    if (props.logToTerminal) {
+      ws.addEventListener("message", (event: MessageEvent) => {
+        if (typeof event.data === "string") {
+          const evt = JSON.parse(event.data);
+          if (evt.method === "Runtime.exceptionThrown") {
+            const params = evt.params as Protocol.Runtime.ExceptionThrownEvent;
+            console.error(
+              "🚨", // cheesy, but it works
+              // maybe we could use color here too.
+              params.exceptionDetails.text,
+              params.exceptionDetails.exception.description
+            );
+          }
+          if (evt.method === "Runtime.consoleAPICalled") {
+            logConsoleMessage(
+              evt.params as Protocol.Runtime.ConsoleAPICalledEvent
+            );
+          }
+        } else {
+          // We should never get here, but who know is 2022...
+          console.error("unrecognised devtools event:", event);
+        }
+      });
+    }
+
+    ws.addEventListener("open", () => {
+      send({ method: "Runtime.enable", id: messageCounterRef.current });
+      // TODO: This doesn't actually work. Must fix.
+      send({ method: "Network.enable", id: messageCounterRef.current++ });
+
+      keepAliveInterval = setInterval(() => {
+        send({
+          method: "Runtime.getIsolateId",
+          id: messageCounterRef.current++,
+        });
+      }, 10_000);
+    });
+
+    ws.on("unexpected-response", () => {
+      console.log("waiting for connection...");
+      /**
+       * This usually means the worker is not "ready" yet
+       * so we'll just retry the connection process
+       */
+      retryRemoteWebSocketConnection();
```

Comment:
Is there an upper limit on this, or retries indefinitely?

Reply:
Indefinitely. which is... fine, I think. It takes a minute or so between retries, and stops the moment you open the browser tab or devtools. We can revisit this if it's a problem.